gmp-tcs-offset/src/main/java/edu/gemini/aspen/gmp/tcsoffset/epics/EpicsTcsOffsetIOC.java

```diff
@@ -0,0 +1,563 @@
+     * @param values: Double list with the current values changed.
+     *
+     */
+
+    public void setTcsInPos(List<Double> values) {
+        for (Double e : values) {
+            synchronized (this) {
+                _tcsIsInPosition = (e == 1.0);
+                if (_isWaitingForInPos)
+                    this.notify();
+            }
+        }
+    }
+
+    /**
+     * This function is associated to <tcsTop>:applyC monitor and
+     * this is called for any update in this monitor. 
+     * @param lStatus: Double list with the current values changed.
+     *
+     */
+
+    public void setTcsStatus(List<Short> lStatus) {
+        for (Short e : lStatus) {
+            synchronized (this) {
+                _tcsState = CARSTATE.getFromInt(e);
+                if (_isWaitingForState)
+                    this.notify();
+            }
+        }
+    }
+
+    private boolean areChannelsInit() {
+        if (_trackingFrameChannel != null  &&  _offsetSizeChannel != null
+            && _angleChannel != null && _virtualTelChannel != null
+            && _tcsApply != null && (!_chLoops.isEmpty())) {
+            return true;
+        }
+
+        return initializeChannels();
+    }
+
+
+    private boolean waitChange(int timeout) {
+        try {
+            this.wait(timeout);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * This function waits X milliseconds until the TCS reaches
+     * the expected value.
+     * @param timeout    : timeout in milliseconds. 
+     * @param expectValue: Expected value before X milliseconds pass. 
+     *
+     * @return           : True, if the expected value was reached before the X milliseconds. 
+     *                   : False, if the expected value was not reach before the X milliseconds. 
+     */
+
+    private boolean waitTcs(int timeout, CARSTATE expectValue) {
+        long rest, t1=0;
+
+        if (expectValue == _tcsState)
+            return true;
+        synchronized (this) {
+            t1 = System.currentTimeMillis();
+            _isWaitingForState = true;
+            waitChange(timeout);
+           _isWaitingForState = false;
+        }
+        rest = System.currentTimeMillis() - t1;
+
+        if (_tcsState != expectValue) {
+            return false;
+        }
+        return true;
+    }
+    /**
+     * This function waits X milliseconds until the TCS reaches
+     * the expected value.
+     * @param timeout    : timeout in milliseconds. 
+     * @param inPosition : Expected value before X milliseconds regarding on TCS in position flag.
+     *
+     * @return           : True, if the expected value was reached before the X milliseconds. 
+     *                   : False, if the expected value was not reach before the X milliseconds. 
+     */
+
+    private boolean waitTcsInPos(int timeout,  Boolean inPosition) {
+        if (inPosition == _tcsIsInPosition)
```

Comment:
Synchronized code is very tricky to get right and the low-level primitives are best avoided if possible.  It might be better to use a `CountDownLatch` or `Semaphore` or something?  I think long ago when I used to do this kind of code in Java though I might have written it something like this:

```java 
private synchronized boolean waitTcsInPos(long timeout, boolean inPosition) {

  long now = System.currentTimeMillis();
  final long end = now + timeout;

  while ((_tcsIsInPosition != inPosition) && (now < end)) {
     try {
       wait(timeout);
     } catch (InterruptedException ex) {
       LOG.info("Interrupted while waiting for TCS in position == " + inPosition);
       break;
     }
     now = System.currentTimeMillis();
  }

  return _tcsIsInPosition == inPosition;

}
```

That way you're always testing the condition with the lock held.  If we have access to fields from synchronized and not synchronized blocks the results will not be predictable.

Reply:
Thank you very much Shane for your comment and for providing an example. 
Regarding the example, the while loop if its condition is met will only be executed once, if the wait(timeout) is completed the now variable it would greater that end (end=now + timeout). Therefore, the logic would be practically the same to put. 
```
private synchronized boolean waitTcsInPos(long timeout, boolean inPosition) {
   if (_tcsInPosition == inPosition)
      return true;
   try {
      wait(timeout);
   } catch (InterruptedException e) { 
      LOG.info("Interrupted while waiting for TCS in position == " + inPosition);
  }
return _tcsIsInPosition == inPosition;
```
However, we could take advantage of your idea and just change the wait(timeout/10) line. Maybe you wanted to put that in. 

On the other hand, I'm more in support of using the synchronized block than the synchronized method, because a synchronized method imposes lock on the entire functionality of the respective method, while the synchronized block is used to lock a lesser number of consecutive statements. However, just for this implementation it doesn't matter and it's not going to get much better. I have changed the code to always use synchronized method.

Reply:
The reason for the loop is that there's no guarantee that your thread was awakened by a notify or that it is the notify that you're expecting.  For that reason it's recommended to always put it in a loop and check the condition.  Also, I notice that I made a mistake (see it's tricky :-) ) in that subsequent wait timeouts should be reduced by however long is left until `end` instead of reusing `timeout`.  `wait(end - now)` maybe.